metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -284,6 +334,120 @@ FinalizedControllerFeatures finalizedFeatures(long epoch) {
+                    break;
+                case MIGRATION:
+                    // This cluster is migrated, but still in dual-write mode
+                    if (zkMigrationEnabled) {
+                        log.debug("{} read a ZkMigrationState of MIGRATION indicating this cluster is being migrated from ZK.", prefix);
+                    } else {
+                        throw new IllegalStateException(
+                                prefix + " read a ZkMigrationState of MIGRATION indicating this cluster is being migrated " +
+                                        "from ZK, but the controller does not have migrations enabled."
+                        );
+                    }
+                    break;
+                case PRE_MIGRATION:
+                    if (!state.preMigrationSupported()) {
+                        // Upgrade case from 3.4. The controller only wrote PRE_MIGRATION during migrations in that version,
+                        // so this needs to complete that migration.
+                        log.info("{} read a ZkMigrationState of PRE_MIGRATION from a ZK migration on Apache Kafka 3.4.", prefix);
+                        if (zkMigrationEnabled) {
+                            recordConsumer.accept(buildRecord(ZkMigrationState.MIGRATION));
+                            log.info("Writing ZkMigrationState of MIGRATION since migration mode is still active on the controller.");
+                        } else {
+                            recordConsumer.accept(buildRecord(ZkMigrationState.MIGRATION));
+                            recordConsumer.accept(buildRecord(ZkMigrationState.POST_MIGRATION));
+                            log.info("Writing ZkMigrationState of POST_MIGRATION since migration mode is not active on the controller.");
+                        }
+                    } else {
+                        log.error("{} read a ZkMigrationState of PRE_MIGRATION indicating this cluster failed during a ZK migration.", prefix);
+                        throw new IllegalStateException("Detected an invalid migration state during startup, cannot continue.");
+                    }
+                    break;
+                default:
+                    throw new IllegalStateException("Unsupported migration state " + state.zkMigrationState());
+            }
+        }
+    }
+
+    /**
+     * Tests if the controller should be preventing metadata updates due to being in the PRE_MIGRATION
+     * state. If the controller does not yet support migrations (before 3.4-IV0), then metadata updates
+     * are allowed in any state. Once the controller has been upgraded to a version that supports
+     * migrations, then this method checks if the controller is in the PRE_MIGRATION state.
+     */
+    boolean inPreMigrationMode(MetadataVersion metadataVersion) {
+        ZkMigrationControlState state = migrationControlState.get();
```

Comment:
This should just be `return state.zkMigrationState() == ZkMigrationState.PRE_MIGRATION;` right? The static configuration isn't related to the state, which should just reflect the metadata log.